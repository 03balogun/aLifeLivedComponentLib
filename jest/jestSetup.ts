/* eslint-disable no-undef */
jest.useFakeTimers('legacy'); // Stops reference error from appearing
