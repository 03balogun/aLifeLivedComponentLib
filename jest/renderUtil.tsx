import React from 'react';
import { ThemeProvider } from '../src/themes/ThemeProvider';
import { render as rtlRender } from '@testing-library/react-native';

function render(ui, { ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <ThemeProvider>{children}</ThemeProvider>;
  }

  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

// re-export everything for convenience
export * from '@testing-library/react-native';

// override render method
export { render };
