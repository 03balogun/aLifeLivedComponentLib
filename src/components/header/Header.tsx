import React from 'react';
import { IconProp } from '@fortawesome/fontawesome-svg-core';
import { Box } from '../../themes/box/Box';
import { SystemButton } from '../../themes/systemButton/SystemButton';
import { Text } from '../../themes/text/Text';
import { Avatar } from '../avatar/Avatar';
import { Icon } from '../icon/Icon';

type HeaderProps = {
  fullName: string;
  avatarUrl: string;
  avatarTestID?: string;
  iconTestID?: string;
  iconName?: IconProp;
  onIconPress?: () => void;
  onAvatarPress: () => void;
};

export const Header: React.FC<HeaderProps> = ({
  fullName,
  avatarUrl,
  avatarTestID,
  iconTestID,
  iconName,
  onIconPress,
  onAvatarPress,
}) => {
  return (
    <Box
      flexDirection="row"
      justifyContent="space-between"
      alignContent="center"
      alignItems="center"
    >
      <Box flexDirection="row" alignContent="center" alignItems="center">
        <SystemButton
          testID={avatarTestID}
          flexDirection="row"
          alignItems="center"
          onPress={onAvatarPress}
        >
          <Avatar
            source={{ uri: avatarUrl, priority: 'high' }}
            iconHeight={40}
            iconWidth={40}
          />
        </SystemButton>
        <Box marginLeft="s" width="75%">
          <Text variant="h4" numberOfLines={1}>
            {fullName}
          </Text>
        </Box>
      </Box>
      {iconName && (
        <SystemButton testID={iconTestID} onPress={onIconPress}>
          <Icon size={24} icon={iconName} color="backgroundContrastMid" />
        </SystemButton>
      )}
    </Box>
  );
};
