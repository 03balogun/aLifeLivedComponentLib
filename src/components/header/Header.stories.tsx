import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { text } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';

// Component
import { Header } from './Header';
import { Box } from '../../themes/box/Box';

storiesOf('Headers', module)
  .addDecorator(Story => Story())
  .add('Header', () => (
    <Box margin="m">
      <Header
        fullName={text('Title', 'John Smith')}
        avatarUrl="https://picsum.photos/id/237/200/300"
        iconName="bookmark"
        onAvatarPress={action('Avatar press')}
        onIconPress={action('Search icon press')}
        avatarTestID="avatar-test-id"
        iconTestID="icon-test-id"
      />
    </Box>
  ));
