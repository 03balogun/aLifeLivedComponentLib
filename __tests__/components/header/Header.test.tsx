import React from 'react';
import { IconProp } from '@fortawesome/fontawesome-svg-core';
import { render, fireEvent } from '../../../jest/renderUtil';

// Component
import { Header } from '../../../src/components/header/Header';

const onIconPressMock = jest.fn();
const onAvatarPressMock = jest.fn();

const setup = (iconName?: IconProp) =>
  render(
    <Header
      fullName="John Smith"
      avatarUrl=""
      avatarTestID="avatar-test-id"
      iconTestID="icon-test-id"
      iconName={iconName}
      onIconPress={onIconPressMock}
      onAvatarPress={onAvatarPressMock}
    />
  );

describe('Given header is rendered', () => {
  test('Should match snapshot', () => {
    const { toJSON } = setup();
    expect(toJSON()).toMatchSnapshot();
  });

  test('Should find welcome message', () => {
    const { getByText } = setup();
    expect(getByText('John Smith')).toBeDefined();
  });

  test('Should find avatar by test id', () => {
    const { getByTestId } = setup();
    expect(getByTestId('avatar-test-id')).toBeDefined();
  });

  describe('Given icon name is provided', () => {
    test('Should find icon by test id', () => {
      const { getByTestId } = setup('search');
      expect(getByTestId('icon-test-id')).toBeDefined();
    });

    describe('Given user presses search icon', () => {
      test('Should fire on search icon press', () => {
        const { getByTestId } = setup('search');
        fireEvent.press(getByTestId('icon-test-id'));
        expect(onIconPressMock).toBeCalledTimes(1);
      });
    });
  });

  describe('Given icon name is NOT provided', () => {
    test('Should NOT render icon', () => {
      const { queryByTestId } = setup();
      expect(queryByTestId('icon-test-id')).toBeNull();
    });
  });

  describe('Given user presses avatar', () => {
    test('Should fire on avatar press', () => {
      const { getByTestId } = setup();
      fireEvent.press(getByTestId('avatar-test-id'));
      expect(onAvatarPressMock).toBeCalledTimes(1);
    });
  });
});
