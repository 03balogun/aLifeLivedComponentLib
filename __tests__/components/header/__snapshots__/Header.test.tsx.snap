// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given header is rendered Should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "row",
        "justifyContent": "space-between",
      },
      Array [
        Object {},
      ],
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "alignContent": "center",
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
        },
        Array [
          Object {},
        ],
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "opacity": 1,
        }
      }
      testID="avatar-test-id"
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#67B6A8",
              "borderRadius": 96,
              "borderWidth": 0,
              "display": "flex",
              "justifyContent": "center",
              "padding": 16,
            },
            Array [
              Object {},
            ],
          ]
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "display": "flex",
            "marginLeft": 8,
            "width": "75%",
          },
          Array [
            Object {},
          ],
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#282828",
              "display": "flex",
              "fontFamily": "SFUIText-Regular",
              "fontSize": 20,
            },
          ]
        }
      >
        John Smith
      </Text>
    </View>
  </View>
</View>
`;
